### ISSUE TITLE
Fit the thermal model

### ISSUE THREAD

**Issue by pattacini** [robotology/icub-firmware-models#36]
Let's proceed with identifying the parameters $R_t$ and $\tau$ robustly by covering a wide range of operating conditions with currents and temperatures (it'll certainly take a while to do so).

As usual, the validation will need to be performed on motors/setup/operating-conditions not seen during fitting.

_Originally posted by @pattacini in https://github.com/robotology/icub-firmware-models/issues/34#issuecomment-1527158675_

**Comment by pattacini** [comment]
We will be employing:
- the thermocouple, to measure the "absolute" starting and ending temperatures;
- the thermocamera, to measure the temporal evolution.

**Comment by mfussi66** [comment]
Today I gathered some meaningful data after some trials, enough for a very first estimation of the model. Here is the data considering the computed dissipated power, and the measured temperature with the camera with the initial value removed.

![image](https://user-images.githubusercontent.com/38140169/235977992-dc738b31-e825-4e62-bc18-b211da8ee0de.png)

I measured the temperature on the wires plastic connector, that would get significantly hotter than the metal cap. Moreover, the temperature oscillates significantly less.

I fitted the model with the training data, with the following result:

![image](https://user-images.githubusercontent.com/38140169/235978899-0fe20406-de40-4917-a0ff-7aa2d15e9286.png)

and validated on the validation data:

![image](https://user-images.githubusercontent.com/38140169/235979068-09625ce1-9150-4ce5-82fb-eed5d12bbcca.png)

At the moment the model differs significantly from what we computed at first:


```
Process model with transfer function:
             Kp                      
  G(s) = ----------                  
          1+Tp1*s                    
                                     
        Kp = 11.981
       Tp1 = 402.55    
```

The results can definitely be improved by fitting with multiple experiments.

**Comment by pattacini** [comment]
Some notes:
- Remember to ensure the model starts off from `0` in validation.
- I would have expected not a perfect match in validation, but a more minor discrepancy, honestly. This leads me to think that there might be aspects we're not catching with our simple thermal model.
- It's tedious to wait for such long test runs, but we also need to make sure that the exponential transition is complete and that we robustly reach for the steady state in our data.

**Comment by mfussi66** [comment]
A second acquisition with the same conditions shows good repeatability and fit of the model, at least in the conditions of:
- 1Kg payload
- Yaw motion at 40deg/s
![1kg](https://user-images.githubusercontent.com/38140169/236176915-89ab2fd8-50cd-4c47-8dbd-7ba1d888a7df.png)


Currently I'm acquiring data with
- 1.5Kg
- No motion
- Longer settling time

so I can see the behaviour in a different scenario, and check if the previous model matches the following one.

**Comment by mfussi66** [comment]
> Currently I'm acquiring data with
> 
> * 1.5Kg
> * No motion
> * Longer settling time (33min acquisition)
> 
> so I can see the behaviour in a different scenario, and check if the previous model matches the following one.

Two acquisitions mentioned above confirm the successful repeatability of the experiment in different conditions.

Here is the resulting plot, with RMSE of 0.6 degC.

![1 5kgno](https://user-images.githubusercontent.com/38140169/236199148-30869211-6f79-47b0-a749-bba34ed28f57.png)


However, we can notice even by the naked eye that there will be  differences in the two models estimated in two different conditions. Summarizing them here:

$$
\frac{\Delta T}{P} (s)= \frac{R_{th}}{\tau s + 1}
$$

||1Kg Yaw|  1.5Kg No yaw|
|---|---|---|
|tau|484.9| 402.6|
|Rth|15.63|10.46|

Maybe this discrepancy could be caused by not reaching exactly steady state conditions in the first case.

**Comment by pattacini** [comment]
The step responses of the two systems above:

<img width="800" alt="image" src="https://user-images.githubusercontent.com/3738070/236204410-46a59718-c29c-47fb-8a70-b749b29f6a0f.png">

**Comment by mfussi66** [comment]
Running the experiment with
- 1Kg
- no motion

Results in a model that gets closer to the one found with 1.5Kg:

||1Kg No Yaw|  1.5Kg No yaw|
|---|---|---|
|$\tau$|**438.25**| 402.6|
|Rth|**12.67**|10.46|

![1 5kg-1Kgno](https://user-images.githubusercontent.com/38140169/236215276-8e6ff3df-d708-43e0-a4ae-7bab65fae57b.png)

We might need a scaling factor in input to the model, which could coincide with the rms value of the input signal (can be computed easily in simulink).

**Comment by mfussi66** [comment]
> Going a bit ahead, here're some nice results from using `Moving RMS`:
> 
> <img alt="image" width="1358" src="https://user-images.githubusercontent.com/3738070/236249809-a647ef6b-0f41-49f1-9383-b4bbe6dacedc.png">
> 
> The model: [**`test_rms.zip`**](https://github.com/robotology/icub-firmware-models/files/11398888/test_rms.zip).

I therefore try to:
- Compute the Id and Iq RMS values of an experiment with varying currents
- Use the computed RMS to fit the model
- Validate the model with a dataset in which the current is kept constant

**Comment by mfussi66** [comment]
After adding the moving RMS block, I tuned its exponential weighing factor to 0.995 empirically, accounting for the fact that decay during repeated motion at 40deg/s should not reach zero.
Then, I exposed directly the computed power via CAN, avoiding extra computations on the host.

Here is the training set taken by moving the yaw with 1Kg payload:

![image](https://user-images.githubusercontent.com/38140169/236527859-3d25d7a4-25d5-4993-a726-f1b5c7f82ea0.png)

The fit on the training data is of course very good (90%):

![image](https://user-images.githubusercontent.com/38140169/236528233-e7d14f36-327a-47cc-9c60-c6b0abbee39a.png)

The model in this acquisition has a significantly higher time constant, but comparable Rth to the previous experiments:

```
        Rth = 14.345                  
       tau = 586.78
```

I wanted to validate the model by getting data with a stopped rotor, but the thermal camera software kept crashing 😓. 

Nonetheless, I got the initial and final temperature (respectively 27.7degC and 53.7 degC), together with the computed power (which is constant in this case, around 2.16W). So we can infer from these parameters at least the Rth for the validation set:

(53.7 degC - 27.7 degC) / 2.16 W = 12.03

12.03 is 17% different from 14.35. 

We can surely acquire even more data to get more robust conclusions, but I think the takeaway is that the RMS block is needed to get more realistic values.

**Comment by pattacini** [comment]
Superb! 🚀 

> I tuned its exponential weighing factor to 0.995 empirically, accounting for the fact that decay during repeated motion at 40deg/s should not reach zero.

Is this tuning general enough for running pretty much in all operational conditions?

Also, I think you can move on and propose the addition of the RMS block to the model via a PR.

**Comment by mfussi66** [comment]
I think it depends on how the input signal varies. Given the application on the wrist at the moment, I think it's safe to keep the decay rate relatively quick to account for the types of sudden movements that the robot can make.
We can explore a range of waves (even synthetically) to verify my suppositions.

**Comment by mfussi66** [comment]
> Also, I think you can move on and propose the addition of the RMS block to the model via a PR.

Landed here: https://github.com/robotology/icub-firmware-models/pull/37

**Comment by mfussi66** [comment]
> We can surely acquire even more data to get more robust conclusions, but I think the takeaway is that the RMS block is needed to get more realistic values.

Further trials today showed that there is still a difference between model and validation set, when the latter is taken with a moving wrist. 

# Constant current

## Same input current
We have a good match when both training and validation set have a constant current, with two very similar amplitudes.

![image](https://user-images.githubusercontent.com/38140169/236839175-0c3cd289-aba7-472c-9467-73fe679bc2a4.png)

![image](https://user-images.githubusercontent.com/38140169/236839049-2fc67838-23fa-49de-a62d-bc2f1390139d.png)

## Different input current

This good match occurs also when validating against current with a different amplitude (as much as the wrist allows):

![image](https://user-images.githubusercontent.com/38140169/236840420-403bbe9b-8b4f-4d99-a75c-bec64f235f0d.png)

![image](https://user-images.githubusercontent.com/38140169/236840827-7fa49449-8b45-4132-a6a8-a4f79ef37db4.png)

In this case the model has: **Rt = 10.2 tau=384**

# Periodic current

Here the datasets were taken by moving the yaw repeatedly:
![image](https://user-images.githubusercontent.com/38140169/236842151-a67c2892-b1e8-4f63-b080-1c1ff3a8d5fc.png)
![image](https://user-images.githubusercontent.com/38140169/236842389-21818acb-3c78-4841-b8e5-8e152c1b9437.png)

In this case the model has: **Rt = 17 tau=495**


# Training with constant current, validation with periodic current

In this scenario the difference is significant, as one can imagine, since the Rt differ a lot. 

![image](https://user-images.githubusercontent.com/38140169/236844858-4d6f5155-c716-4598-b125-6c6442c8a41c.png)


A significant role in this difference is probably played by the 3 factors:
- downsampling of the current
- Rms value computed in a window of time
- Current waveform

**Comment by pattacini** [comment]
We agreed to:
- Update the dictionary to rely on the params that came out from identifying against the constant current.
- Run the RMS within the FOC to gauge the effect of sampling.

**Comment by pattacini** [comment]
>- Run the RMS within the FOC to gauge the effect of sampling.

Once in this condition, we may need to retune the Moving RMS params to get a more stable estimation of the power.

**Comment by mfussi66** [comment]
I created a bespoke model to play around with different sampling times and weighing factor $\lambda$ of the moving RMS block.

The currents Iq and Id belong to the experiment with moving wrist, acquired at 10Hz and resampled (through interpolation) at 26.6KHz to emulate the FOC rate. Having them, I computed the dissipated power

![MATLAB_mdCuS7TtlY](https://github.com/robotology/icub-firmware-models/assets/38140169/6b75be86-78d1-43e8-8031-e6da57fc1048)

I tried the following combinations:

|Ts| $\lambda$ |
|---|---|
| 26.6 KHz | 1 (true RMS at t->inf) | 
| 26.6 KHz | 0.995 | 
| 26.6 KHz | 0.99999 | 
| 100 Hz | 0.99 | 
| 100 Hz | 0.995 | 
|100 Hz | 0.999 | 
|10 Hz | 0.995 | 

The resulting plot at the moment is a bit chaotic:

![eval](https://github.com/robotology/icub-firmware-models/assets/38140169/e7820b03-ceca-4281-be71-c21e394b195e)

So let's zoom on two relevant waveforms:

### Step signal
![eval_detail](https://github.com/robotology/icub-firmware-models/assets/38140169/31f40623-e512-434c-9d49-22296a829c21)

### Periodic signal
![eval_detail_periodic](https://github.com/robotology/icub-firmware-models/assets/38140169/6ff4939e-9dd7-4300-beff-ddc5dbdccef6)

We can see that:
- At the FOC rate, $\lambda$ needs to be significantly increased compared to when running it at 1e-2
- Using the same $\lambda$ at 100Hz or 10 Hz changes significantly the settling time and the difference compared to the true rms

**Comment by pattacini** [comment]
Nice statistics 👍🏻 
We can comment on them F2F.

Just one important note though.

>The currents Iq and Id belong to the experiment with moving wrist, acquired at 10Hz and resampled (through interpolation) at 26.6KHz to emulate the FOC rate.

I don't think this works ok. Upsampling a signal doesn't increase its real frequency content.

**Comment by mfussi66** [comment]
> I don't think this works ok. Upsampling a signal doesn't increase its real frequency content. 

Of course, the resampling was just to establish a baseline and how the same lambda changes the results.

**Comment by mfussi66** [comment]
Me and @pattacini checked F2F the plots in the simulation above and agreed to keep the computation of the RMS in the FOC with $\lambda = 1$ , relying on the fact that the time constant of the temperature is slow enough to allow for precise RMS evaluation.

**Comment by mfussi66** [comment]
Here I ran the same motion with lambda=1 in the FOC, to qualitatively show how fast is the moving RMS response is. It takes 500 seconds to go from 0.3W to 0.1W, which is a time constant that is higher than the estimated thermal model, so it might affect the temperature estimation.

![taurms](https://github.com/robotology/icub-firmware-models/assets/38140169/c0459c17-91e4-40fc-82eb-714b478965cf)


Nonetheless, I can try some acquisitions.

**Comment by pattacini** [comment]
> a time constant that is higher than the estimated thermal model, so it might affect the temperature estimation.

I wouldn't be worried in this case as there's a temperature drop here rather than an increase.

**Comment by pattacini** [comment]
Hi @mfussi66 

I've moved the RMS computation inside the FOC via 2b1c93cc3b739fb1b6945ff3a9a3fadbe4f02eb3.

**Comment by mfussi66** [comment]
While getting more data to verify the fitting accuracy, I noticed that the evaluation of the RMS currents did not match my expectations in some conditions.
These conditions involve switching the current waveform, e.g. from:
1. variable (due to the moving wrist)
2.  to constant
3. and back to variable (due to the moving wrist)

Since $\lambda=1$, the RMS inside the code gives equal weight to all the samples, and simply accumulates them.  We expect RMS(point 1) == RMS (point 3), but since we switched to a constant waveform halfway, the accumulator is now polluted. 

I made a Simulink model to test this hypothesis, in which I periodically switch from constant to variable current (sine wave + noise), and it seems to confirm it:

![rms_check_sequence](https://github.com/robotology/icub-firmware-models/assets/38140169/8ac1e5ce-5a07-4a16-9ea2-50519bb26df9)

With $\lambda=1$ the RMS value is never able to reach $1/\sqrt{2}$ as expected. Even worse, its derivative goes to zero and the value settles to a lower value than expected:

![rms_check_sequence_deriv](https://github.com/robotology/icub-firmware-models/assets/38140169/9c4d24d3-371d-40fd-bc90-361e52c846f2)

This is a consideration that [in this previous comment and test](https://github.com/robotology/icub-firmware-models/issues/36#issuecomment-1542445469) it was probably much harder to notice.

My suggestion is to keep $\lambda < 1$, such as 0.9995, taking into account the limitations of `single` datatype.

**Comment by pattacini** [comment]
Somehow, I cannot click on images for inspection.

**Comment by mfussi66** [comment]
Speaking with @pattacini , we agreed to keep the lambda=0.9995 and also acquire datasets at different speeds, to account for the fact that a constant current regime might be a limit condition.

**Comment by mfussi66** [comment]
I got three datasets, trying to vary the speed as much as the wrist allows (but I can take more for better cross validation).
I fitted the model with one and validated with the other two, with very good results: 

![image](https://github.com/robotology/icub-firmware-models/assets/38140169/3577f428-2695-4dd2-aef4-e64607d18ba7)

In both cases the RMSE is very good < 0.5 degC.

Though the estimated parameters are significantly different compared to the constant current case:

Rt = 16
tau = 797.5

and they will overestimate the temperature (which could be good, since it would be a cautious measure). This can be seen in the lines in magenta, in which the estimated temperature is 4 degC higher than the measured one.

**Comment by pattacini** [comment]
Nice result!

Let's put the case of constant current aside for a moment.
We should:
- Consider which **equivalent critical temperature** to account for to keep the windings safe (it'll be lower than that one in the datasheet). Probably, run a few tests for sanity check.
- Make a PR to update Rt, tau, lambda, Tcrit.

**Comment by mfussi66** [comment]
> Instead, to complete this issue, we do just need to update the values of Rt, tau, lambda (and Tcrit?).

PR here: https://github.com/robotology/icub-firmware-models/pull/43
